app/stac_api/serializers.py

```diff
@@ -389,6 +403,61 @@ def to_representation(self, instance):
         )
         return representation
 
+    def _validate_general_url_pattern(self, url):
+        try:
+            validator = URLValidator()
+            validator(url)
+        except exceptions.ValidationError as exc:
+            errors = {'href': _('Invalid URL provided')}
+            raise serializers.ValidationError(code='payload', detail=errors) from exc
+
+    def _validate_configured_url_pattern(self, url):
+        # get the collection regex. if there's none, allow all urls
+        collection_regex = self.collection.external_asset_pattern or r".*"
+
+        if not re.search(collection_regex, url):
+            errors = {'href': _("Invalid URL provided. It doesn't match the collection pattern")}
+            raise serializers.ValidationError(code='payload', detail=errors)
+
+    def _validate_reachability(self, url):
+        unreachable_error = {'href': _('Provided URL is unreachable')}
+        try:
+            response = requests.head(url, timeout=settings.EXTERNAL_URL_REACHABLE_TIMEOUT)
+            if response.status_code > 400:
+                raise serializers.ValidationError(code='payload', detail=unreachable_error)
+        except requests.Timeout as exc:
+            errors = {'href': _('Checking href URL resulted in timeout')}
+            raise serializers.ValidationError(code='payload', detail=errors) from exc
+        except requests.ConnectionError as exc:
+            raise serializers.ValidationError(code='payload', detail=unreachable_error) from exc
+
+    def _validate_href_url(self, url):
+        self._validate_general_url_pattern(url)
+        self._validate_configured_url_pattern(url)
+        self._validate_reachability(url)
+
+    def _validate_href_field(self, attrs):
+        """Only allow the href field if the collection allows for external assets
+
+        Raise an exception, this replicates the previous behaviour when href
+        was always read_only
+        """
+        if 'file' in attrs:
+            if self.collection:
+                collection = self.collection
+            else:
+                raise Exception("Implementation error")
+
+            if not collection.allow_external_assets:
+                errors = {'href': _("Found read-only property in payload")}
+                raise serializers.ValidationError(code="payload", detail=errors)
+
+            self._validate_href_url(attrs['file'])
+
+    def validate(self, attrs):
+        self._validate_href_field(attrs)
+        return super().validate(attrs)
```

Comment:
One more thing that I just realized: I dunno what will be logged in this part by default, but it's probably a section that we want to have a bit more verbosity regarding logging to be able to debug issues